File: src/main/java/model/PoliformatFile.java
package model;

import com.squareup.moshi.Json;
import model.json.adapter.Clean;

import java.net.URL;
import java.nio.file.Path;

public class PoliformatFile {

    @Json(name = "author")
    private String author;
    @Json(name = "authorId")
    private String authorId;
    @Json(name = "container")
    private Path container;
    @Json(name = "copyrightAlert")
    private String copyrightAlert;
    @Json(name = "description")
    private String description;
    @Json(name = "endDate")
    private String endDate;
    @Json(name = "fromDate")
    private String fromDate;
    @Json(name = "modifiedDate")
    private String modifiedDate;
    @Json(name = "numChildren")
    private Integer numChildren;
    @Json(name = "quota")
    private String quota;
    @Json(name = "size")
    private Integer size;
    @Clean
    @Json(name = "title")
    private String title;
    @Json(name = "type")
    private String type;
    @Json(name = "url")
    private URL url;
    @Json(name = "usage")
    private String usage;
    @Json(name = "hidden")
    private Boolean hidden;
    @Json(name = "visible")
    private Boolean visible;
    @Json(name = "entityReference")
    private String entityReference;
    @Json(name = "entityURL")
    private String entityURL;
    @Json(name = "entityTitle")
    private String entityTitle;

    public String getParentUrl() {
        String urlString = url.toString();
        int index;

        if (urlString.endsWith("/")) {
            index = urlString.substring(0, urlString.length() - 1).lastIndexOf('/');
        } else {
            index = urlString.lastIndexOf('/');
        }

        return url.toString().substring(0, index + 1);
    }

    public String getAuthor() {
        return author;
    }

    public String getAuthorId() {
        return authorId;
    }

    public Path getContainer() {
        return container;
    }

    public String getCopyrightAlert() {
        return copyrightAlert;
    }

    public String getDescription() {
        return description;
    }

    public String getEndDate() {
        return endDate;
    }

    public String getFromDate() {
        return fromDate;
    }

    public String getModifiedDate() {
        return modifiedDate;
    }

    public Integer getNumChildren() {
        return numChildren;
    }

    public String getQuota() {
        return quota;
    }

    public Integer getSize() {
        return size;
    }

    public String getTitle() {
        return title;
    }

    public String getType() {
        return type;
    }

    public URL getUrl() {
        return url;
    }

    public String getUsage() {
        return usage;
    }

    public Boolean getHidden() {
        return hidden;
    }

    public Boolean getVisible() {
        return visible;
    }

    public String getEntityReference() {
        return entityReference;
    }

    public String getEntityURL() {
        return entityURL;
    }

    public String getEntityTitle() {
        return entityTitle;
    }

}

File: src/main/java/model/PoliformatFile.kt
package model

import com.squareup.moshi.Json
import model.json.adapter.Clean

import java.net.URL
import java.nio.file.Path

class PoliformatFile(
        @Json(name = "author")
        val author: String? = null,
        @Json(name = "authorId")
        val authorId: String? = null,
        @Json(name = "container")
        val container: Path? = null,
        @Json(name = "copyrightAlert")
        val copyrightAlert: String? = null,
        @Json(name = "description")
        val description: String? = null,
        @Json(name = "endDate")
        val endDate: String? = null,
        @Json(name = "fromDate")
        val fromDate: String? = null,
        @Json(name = "modifiedDate")
        val modifiedDate: String? = null,
        @Json(name = "numChildren")
        val numChildren: Int? = null,
        @Json(name = "quota")
        val quota: String? = null,
        @Json(name = "size")
        val size: Int? = null,
        @Clean
        @Json(name = "title")
        val title: String? = null,
        @Json(name = "type")
        val type: String? = null,
        @Json(name = "url")
        val url: URL,
        @Json(name = "usage")
        val usage: String? = null,
        @Json(name = "hidden")
        val hidden: Boolean? = null,
        @Json(name = "visible")
        val visible: Boolean? = null,
        @Json(name = "entityReference")
        val entityReference: String? = null,
        @Json(name = "entityURL")
        val entityURL: String? = null,
        @Json(name = "entityTitle")
        val entityTitle: String? = null
) {
    val parentUrl: String by lazy {
        val urlString = url.toString()

        val index = if (isFolder) {
            urlString.substring(0, urlString.length - 1).lastIndexOf('/')
        } else {
            urlString.lastIndexOf('/')
        }

        url.toString().substring(0, index + 1)
    }

    val isFolder = type == "collection"

    override fun equals(other: Any?): Boolean = other is PoliformatFile && this.title == other.title

    override fun toString() = title + " - " + url
}

